addons/gamification/models/res_users.py

```diff
@@ -42,35 +64,29 @@ def _get_user_badge_level(self):
     @api.model_create_multi
     def create(self, values_list):
         res = super(Users, self).create(values_list)
-
-        karma_trackings = []
-        for user in res:
-            if user.karma:
-                karma_trackings.append({'user_id': user.id, 'old_value': 0, 'new_value': user.karma})
+        tracking = self.env['gamification.karma.tracking']
+
+        karma_trackings = [
+            tracking.create_tracking(
+                user.id, 0, vals.get('karma', 0), reason=_('User Creation'))
+            for user, vals in zip(res, values_list)
+            if vals.get('karma', 0)
+        ]
         if karma_trackings:
-            self.env['gamification.karma.tracking'].sudo().create(karma_trackings)
+            tracking.sudo().create(karma_trackings)
 
-        res._recompute_rank()
         return res
 
-    def write(self, vals):
-        karma_trackings = []
-        if 'karma' in vals:
-            for user in self:
-                if user.karma != vals['karma']:
-                    karma_trackings.append({'user_id': user.id, 'old_value': user.karma, 'new_value': vals['karma']})
-
-        result = super(Users, self).write(vals)
-
-        if karma_trackings:
-            self.env['gamification.karma.tracking'].sudo().create(karma_trackings)
-        if 'karma' in vals:
-            self._recompute_rank()
-        return result
-
-    def add_karma(self, karma):
-        for user in self:
-            user.karma += karma
+    def add_karma(self, karma, source=None, reason=''):
```

Comment:
This method is public, it means that I can call it in RPC to give me infinite karma ?
Do you really need it to be public ?

Reply:
this is never called in RPC, so no reason to keep it public